### PR TITLE
feat(category): add support of categories

### PR DESCRIPTION
**Pull request by seeruyy**
Added category resource
- Create category
- Update category
- Read category
- Delete category

LOG-11845

### CODE REVIEW

File: logdna/resource_category.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+  "github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceCategoryCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+  var diags diag.Diagnostics
+  pc := m.(*providerConfig)  
+
+  categoryType := d.Get("type").(string)
```
Comment (darinspivey) [inline logdna/resource_category.go:17]: Creating a resource should allow for `name`, but it's not sent here.  I see that `CreateRequestBody` is filling in `name`, but not `type`.  That's super confusing since name is required, IMO.  It makes it look like things are missing here.

I would move the above line to `CreateRequestBody`, and then to supply it below, pull it from the populated `category` structure below.

File: logdna/request_types.go [hunk 2]
```diff
@@ -77,6 +82,17 @@ func (alert *alertRequest) CreateRequestBody(d *schema.ResourceData) diag.Diagno
 	return diags
 }
 
+func (cateroty *categoryRequest) CreateRequestBody(d *schema.ResourceData) diag.Diagnostics {
+	// This function pulls from the schema in preparation to JSON marshal
+	var diags diag.Diagnostics
+
+	// Scalars
+	cateroty.Name = d.Get("name").(string)
+	// cateroty.Type = d.Get("type").(string)
```
Comment (darinspivey) [inline logdna/request_types.go:91]: I think this `Type` should be set here.  See other comment.

File: logdna/request_types.go [hunk 3]
```diff
@@ -77,6 +82,17 @@ func (alert *alertRequest) CreateRequestBody(d *schema.ResourceData) diag.Diagno
 	return diags
 }
 
+func (cateroty *categoryRequest) CreateRequestBody(d *schema.ResourceData) diag.Diagnostics {
+	// This function pulls from the schema in preparation to JSON marshal
+	var diags diag.Diagnostics
+
+	// Scalars
```
Comment (darinspivey) [inline logdna/request_types.go:89]: This comment isn't really needed since there are no things like lists to separate actions needed in this function.  It's obvious what it's doing.

File: logdna/resource_alert_test.go [hunk 4]
```diff
@@ -154,19 +154,19 @@ func TestAlert_ErrorsSlackChannel(t *testing.T) {
 func TestAlert_ErrorsWebhookChannel(t *testing.T) {
 	btArgs := map[string]map[string]string{"webhook": cloneDefaults(chnlDefaults["webhook"])}
 	btArgs["webhook"]["bodytemplate"] = `"{\"test\": }"`
-	btCfgE := fmtTestConfigResource("alert", "new", nilLst, alertDefaults, btArgs)
+	btCfgE := fmtTestConfigResource("alert", "new", nilLst, alertDefaults, btArgs, nilLst)
 
 	mdArgs := map[string]map[string]string{"webhook": cloneDefaults(chnlDefaults["webhook"])}
 	mdArgs["webhook"]["method"] = `"false"`
-	mdCfgE := fmtTestConfigResource("alert", "new", nilLst, alertDefaults, mdArgs)
+	mdCfgE := fmtTestConfigResource("alert", "new", nilLst, alertDefaults, mdArgs, nilLst)
```
Comment (darinspivey) [inline logdna/resource_alert_test.go:161]: Adding all of these `nilList` because the function signature changed seems really painful and the parameter list is too large at 6 params, IMO.  It's a bigger change, but we should be using a config structure for this.  Then, we could easily default values.

That's the problem with these test functions being abstracted.  It gets really messy trying to make it work for everything.

If you're up to it, make that a new commit.  Create a `testResource{}` structure and add all the things.  `fmtTestConfigRequest` can then be called with that structure and only worry about formatting the properties that are defined.

Fine to create a card to fix later, cause I think we need to get this easier to use before it becomes out of hand.

File: logdna/resource_category.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    DeleteContext: resourceCategoryDelete,
+    Schema: map[string]*schema.Schema{
+      "name": {
+        Type:     schema.TypeString,
+        Required: true,
+      },
+      "type": {
+        Type:     schema.TypeString,
+        Optional: true,
+        Default:  "views",
```
Comment (darinspivey) [inline logdna/resource_category.go:161]: Have you considered setting up an `enum` for this?  I'm not sure what the validation is on the server side--if we have such validation, then not validating here is the choice.  BUT, if there's nothing on the server side, we need to make sure that these aren't bad values.

We do that here for views. https://github.com/logdna/terraform-provider-logdna/blob/c87a6cc7be8fbf70ee2531923580702d08dc80e7/logdna/resource_view.go#L16-L21

File: examples/category.tf [hunk 6]
```diff
@@ -0,0 +1,8 @@
+provider "logdna" {
```
Comment (darinspivey) [inline examples/category.tf:1]: Don't for get documentation in the `docs` directory too.  Just examples isn't enough.  At a minimum, there should be something in `docs/resources`

File: logdna/resource_category.go (hunk 5 above)
Comment (seeruyy) [reply logdna/resource_category.go:161]: @darinspivey We do validate all the request on the server side and type is a enum with a strict list of options. I wanted to add enum validation here as well but I couldn't fund how to do that, I checked terraform schema source code and didn't find any enum option for a schema.

File: logdna/resource_category.go (hunk 1 above)
Comment (seeruyy) [reply logdna/resource_category.go:17]: @darinspivey The type is not part of a request body, that why I set just name in `CreateRequestBody`, type I need just to generate a url. I wanted to avoid sending type field in request body.

File: logdna/resource_alert_test.go (hunk 4 above)
Comment (seeruyy) [reply logdna/resource_alert_test.go:161]: It makes sense. I'll work on this refactoring in a separate story

File: logdna/resource_category.go (hunk 5 above)
Comment (darinspivey) [reply logdna/resource_category.go:161]: If it's server-side validation we'd want to specifically not do that here, IMO.  So you're good.  Might be nice to have a test that rejects a bad value from the server, though.

File: logdna/resource_category.go (hunk 1 above)
Comment (darinspivey) [reply logdna/resource_category.go:17]: Hmm, I get it, but that's confusing since the TF schema has `type` and `name` alongside each other.  Add a comment to clarify and delete the commented out `Type` in `request_types.go`

File: logdna/common_test.go [hunk 7]
```diff
@@ -110,7 +114,7 @@ func fmtProviderBlock(args ...string) string {
 	return fmt.Sprintf(tmplPc, pcCfg)
 }
 
-func fmtResourceBlock(objTyp, rsName string, rsArgs map[string]string, chArgs map[string]map[string]string) string {
+func fmtResourceBlock(objTyp, rsName string, rsArgs map[string]string, chArgs map[string]map[string]string, dependencies []string) string {
```
Comment (mdeltito) [inline logdna/common_test.go:117]: this got a little dicey with all the areas of code it impacts - we should consider refactoring this, not necessarily here in this PR

File: logdna/resource_category_test.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    Steps: []resource.TestStep{
+      {
+        Config: fmtTestConfigResource("category", "new", nilLst, catArgs, nilOpt, nilLst),
+        ExpectError: regexp.MustCompile("The argument \"name\" is required, but no definition was found."),
+      },
+    },
+  })
+}
+
+func TestCategory_Basic(t *testing.T) {
```
Comment (mdeltito) [inline logdna/resource_category_test.go:46]: it looks like we're missing a test case for an update here. 

here's one example but there's other too https://github.com/logdna/terraform-provider-logdna/blob/c87a6cc7be8fbf70ee2531923580702d08dc80e7/logdna/resource_alert_test.go#L226-L230

iirc the `delete` will be tested implicitly by the testing framework (it automatically destroys resources after each test by default)

**Review by mdeltito — CHANGES_REQUESTED** [review]

File: go.mod [hunk 9]
```diff
@@ -3,6 +3,13 @@ module github.com/logdna/terraform-provider-logdna
...
 require (
-	github.com/hashicorp/terraform-plugin-sdk/v2 v2.7.0
+	cloud.google.com/go v0.61.0 // indirect
+	github.com/aws/aws-sdk-go v1.25.3 // indirect
+	github.com/golang/snappy v0.0.1 // indirect
+	github.com/hashicorp/go-getter v1.5.3 // indirect
+	github.com/hashicorp/terraform-plugin-go v0.8.0 // indirect
+	github.com/hashicorp/terraform-plugin-sdk/v2 v2.13.0
+	github.com/mitchellh/cli v1.1.2 // indirect
+	github.com/pierrec/lz4 v2.0.5+incompatible // indirect
```
[the remark this answers is not in the document]
Comment (darinspivey) [reply go.mod:13]: Well, they're all `indirect` which I assumed were deps of the new sdk somehow?

[the remark this answers is not in the document]
Comment (mdeltito) [reply go.mod:13]: I don't think so, you ran into the same thing I think here:

https://github.com/logdna/terraform-provider-logdna/pull/16#discussion_r639888883

(under my review, first file/comment)

[the remark this answers is not in the document]
Comment (seeruyy) [reply go.mod:13]: @mdeltito After upgrade the `sdk` I followed the go.mod suggestions to install additional dependencies to be able to build a binary.

File: logdna/common_test.go (hunk 7 above)
Comment (seeruyy) [reply logdna/common_test.go:117]: @mdeltito yes, it makes sense, @darinspivey also mentioned this point, I'll work on the refactoring in a separate story.

File: go.mod (hunk 9 above)
[the remark this answers is not in the document]
Comment (jorgebay) [reply go.mod:13]: I agree with @mdeltito, the result of `go mod tidy` should be the `go.mod` that goes into source control.

Sadly, [it looks like there's no validation](https://github.com/golang/go/issues/27005) that we could add to CI.

**Review by jorgebay — COMMENTED** [review]
It's looking good to me, besides the doc & coverage remarks already mentioned by Darin and Mike.

File: logdna/resource_category_test.go (hunk 8 above)
Comment (seeruyy) [reply logdna/resource_category_test.go:46]: @mdeltito @jorgebay The name of both test cases is the same, the first case tests create action and the second one tests update action. Overall this case tests the entire lifecycle of a CRUD interface.
Screen Shot 2022-04-19 at 11 21 42 PM
Screen Shot 2022-04-19 at 11 23 05 PM

Comment (darinspivey) [reply logdna/resource_category_test.go:46]: Got it, @seeruyy, thanks.  I do think there's a few style things we could do better on which may help clear up confusion.  Couple things here:

* It's a terrible pattern that we've started using `new` for the resource names.  That is a reserved word in some languages and just doesn't describe the resource.  Even something like `new_thing` would be better.  I'd like to start naming things for what they are and break that pattern (no, we don't have to retroactively fix things as long as new code does better).

* We should rely on heavily using `t.Run` ([like this](https://github.com/logdna/terraform-provider-logdna/blob/529d2fa80335b854fa4aefa67a46cb04bb582851/logdna/request_test.go#L41-L46)) to give proper descriptions and subtests.  The confusion on this test could be cleared up by 2 simple subtests--the first with a description of "creating blah blah" and the second of "updating the same blah blah".

These seem like nits, but we'd do ourselves a favor by improving just a bit so that we don't hurt the readability and intuitiveness of the code.  The tests have already gotten pretty abstracted and can be hard to follow, IMO.

Comment (mdeltito) [reply logdna/resource_category_test.go:46]: > We should rely on heavily using t.Run ([like this](https://github.com/logdna/terraform-provider-logdna/blob/529d2fa80335b854fa4aefa67a46cb04bb582851/logdna/request_test.go#L41-L46)) to give proper descriptions and subtests

I don't think that is true for tests that are using the terraform sdk/test suite? I think the confusion is just that there's some "magic" going on via the sdk which makes it unclear (especially upon picking this project up again) whether we're missing assertions or not.

Comment (darinspivey) [reply logdna/resource_category_test.go:46]: Sorry, I didn't mean to start a big debate.  Yes, there's magic with that test, but it seems that it could be easily wrapped within a child test with a description.  Not having descriptions in the test title or in the various assertions is sucky, IMO.

**Review by jorgebay — DISMISSED** [review]
lgtm!

**Review by mdeltito — APPROVED** [review]
✨ nice, thanks for this

**Review by darinspivey — APPROVED** [review]

**Review by jorgebay — APPROVED** [review]
:shipit:
